File: bindings/java/iota-wallet-java/src/main/java/org/iota/Wallet.java
// Copyright 2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package org.iota;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;
import org.iota.api.GsonSingleton;
import org.iota.api.NativeApi;
import org.iota.types.*;
import org.iota.types.exceptions.WalletException;
import org.iota.types.ids.account.AccountIdentifier;
import org.iota.types.ids.account.AccountIndex;

public class Wallet extends NativeApi {

    public Wallet(WalletConfig config) {
        super(config);
    }

    // Account manager APIs

    public AccountHandle createAccount(String alias) throws WalletException {
        JsonObject o = new JsonObject();
        o.addProperty("alias", alias);

        Account a = GsonSingleton.getInstance().fromJson(callBaseApi(new ClientCommand("CreateAccount", o)), Account.class);
        AccountHandle handle = new AccountHandle(this, new AccountIndex(a.getIndex()));

        return handle;
    }

    public AccountHandle getAccount(AccountIdentifier accountIdentifier) throws WalletException {
        Account a =  GsonSingleton.getInstance().fromJson(callBaseApi(new ClientCommand("GetAccount", GsonSingleton.getInstance().toJsonTree(accountIdentifier))), Account.class);
        AccountHandle handle = new AccountHandle(this, new AccountIndex(a.getIndex()));

        return handle;
    }

    public AccountHandle[] getAccounts() throws WalletException {
        JsonArray responsePayload = (JsonArray) callBaseApi(new ClientCommand("GetAccounts"));

        AccountHandle[] accountHandles = new AccountHandle[responsePayload.size()];
        for (int i = 0; i < responsePayload.size(); i++)
            accountHandles[i] = new AccountHandle(this, new AccountIndex(GsonSingleton.getInstance().fromJson(responsePayload.get(i).getAsJsonObject(), Account.class).getIndex()));

        return accountHandles;
    }


    public void backup(String destination, String password) throws WalletException {
        JsonObject o = new JsonObject();
        o.addProperty("destination", destination);
        o.addProperty("password", password);

        callBaseApi(new ClientCommand("Backup", o));
    }

    public void changeStrongholdPassword(String currentPassword, String newPassword) throws WalletException {
        JsonObject o = new JsonObject();
        o.addProperty("currentPassword", currentPassword);
        o.addProperty("newPassword", newPassword);

        callBaseApi(new ClientCommand("ChangeStrongholdPassword", o));
    }

    public void clearStrongholdPassword() throws WalletException {
        callBaseApi(new ClientCommand("ClearStrongholdPassword"));
    }

    public boolean isStrongholdPasswordAvailable() throws WalletException {
        return callBaseApi(new ClientCommand("IsStrongholdPasswordAvailable")).getAsBoolean();
    }

    public void recoverAccounts(int accountStartIndex, int accountGapLimit, int addressGapLimit, SyncOptions syncOptions) throws WalletException {
        JsonObject o = new JsonObject();
        o.addProperty("accountStartIndex", accountStartIndex);
        o.addProperty("accountGapLimit", accountGapLimit);
        o.addProperty("addressGapLimit", addressGapLimit);
        o.add("syncOptions", GsonSingleton.getInstance().toJsonTree(syncOptions));

        callBaseApi(new ClientCommand("RecoverAccounts", o));
    }

    public void restoreBackup(String source, String password) throws WalletException {
        JsonObject o = new JsonObject();
        o.addProperty("source", source);
        o.addProperty("password", password);

        callBaseApi(new ClientCommand("RestoreBackup", o));
    }

    public void removeLatestAccount() throws WalletException {
        callBaseApi(new ClientCommand("RemoveLatestAccount"));
    }

    public String generateMnemonic() throws WalletException {
        return callBaseApi(new ClientCommand("GenerateMnemonic")).getAsString();
    }

    public String verifyMnemonic(String mnemonic) throws WalletException {
        JsonPrimitive p = new JsonPrimitive(mnemonic);
        return callBaseApi(new ClientCommand("VerifyMnemonic", p)).getAsString();
    }

    public String setClientOptions(ClientConfig config) throws WalletException {
        return callBaseApi(new ClientCommand("SetClientOptions", GsonSingleton.getInstance().toJsonTree(config))).getAsString();
    }

    public LedgerNanoStatus getLedgerNanoStatus() throws WalletException {
        return GsonSingleton.getInstance().fromJson(callBaseApi(new ClientCommand("GetLedgerNanoStatus")), LedgerNanoStatus.class);
    }

    public JsonObject getNodeInfo(String url, NodeAuth auth) throws WalletException {
        JsonObject p = new JsonObject();
        p.addProperty("url", url);
        p.add("auth", GsonSingleton.getInstance().toJsonTree(auth));

        return (JsonObject) callBaseApi(new ClientCommand("GetNodeInfo", p));
    }

    public void setStrongholdPassword(String password) throws WalletException {
        JsonPrimitive p = new JsonPrimitive(password);
        callBaseApi(new ClientCommand("SetStrongholdPassword", p));
    }

    public void setStrongholdPassword(int interval) throws WalletException {
        JsonPrimitive p = new JsonPrimitive(interval);
        callBaseApi(new ClientCommand("SetStrongholdPasswordClearInterval", p));
    }

    public void storeMnemonic(String mnemonic) throws WalletException {
        JsonPrimitive p = new JsonPrimitive(mnemonic);
        callBaseApi(new ClientCommand("StoreMnemonic", p));
    }

    public void startBackgroundSync(SyncOptions options, int intervalInMilliseconds) throws WalletException {
        JsonObject o = new JsonObject();
        o.add("options", GsonSingleton.getInstance().toJsonTree(options));
        o.addProperty("intervalInMilliseconds", intervalInMilliseconds);

        callBaseApi(new ClientCommand("StartBackgroundSync", o));
    }

    public void stopBackgroundSync(SyncOptions options, int intervalInMilliseconds) throws WalletException {
        callBaseApi(new ClientCommand("StopBackgroundSync"));
    }

    public void emitTestEvent(JsonElement event) throws WalletException {
        callBaseApi(new ClientCommand("EmitTestEvent", event));
    }

    public String bech32ToHex(String bech32) throws WalletException {
        return callBaseApi(new ClientCommand("Bech32ToHex", new JsonPrimitive(bech32))).getAsString();
    }

    public String hexToBech32(String hex, String bech32Hrp) throws WalletException {
        JsonObject p = new JsonObject();
        p.addProperty("hex", hex);
        p.addProperty("bech32Hrp", bech32Hrp);

        return callBaseApi(new ClientCommand("HexToBech32", p)).getAsString();
    }

}

File: bindings/python/iota_wallet/wallet.py
import iota_wallet
from iota_wallet.common import send_message_routine
from iota_wallet.account import Account
from json import loads, dumps


class IotaWallet():
    def __init__(self, storage_path='./walletdb', client_options=None, coin_type=None, secret_manager=None):
        """Initialize the IOTA Wallet.
        """

        # Setup the options
        options = {'storagePath': storage_path}
        if client_options:
            options['clientOptions'] = client_options
        if coin_type:
            options['coinType'] = coin_type
        if secret_manager:
            options['secretManager'] = secret_manager

        options = dumps(options)

        # Create the message handler
        self.handle = iota_wallet.create_message_handler(options)

    def get_handle(self):
        return self.handle

    def create_account(self, alias=None, bech32_hrp=None):
        """Create a new account
        """
        return self._send_cmd_routine(
            'createAccount', {
                'alias': self.__return_str_or_none(alias),
                'bech32Hrp': self.__return_str_or_none(bech32_hrp),
            }
        )

    def get_account(self, alias_index):
        """Get the account instance
        """
        return Account(alias_index, self.handle)

    @send_message_routine
    def _send_cmd_routine(self, cmd, payload=None):
        message = {
            'cmd': cmd
        }
        if payload:
            message['payload'] = payload
        return message

    def get_account_data(self, alias_index):
        """Get account data
        """
        return self._send_cmd_routine(
            'getAccount', {
                'accountId': alias_index}
        )

    def get_accounts(self):
        """Get accounts
        """
        return self._send_cmd_routine(
            'getAccounts',
        )

    def backup(self, destination, password):
        """Backup storage.
        """
        return self._send_cmd_routine(
            'backup', {
                'destination': destination,
                'password': password
            }
        )

    def bech32_to_hex(self, bech32_address):
        """Transforms a bech32 encoded address to hex
        """
        return self._send_cmd_routine(
            'bech32ToHex', {
                'bech32Address': bech32_address,
            }
        )

    def change_stronghold_password(self, password):
        """Change stronghold password.
        """
        return self._send_cmd_routine(
            'changeStrongholdPassword', {
                'currentPassword': password,
                'newPassword': password
            }
        )

    def clear_stronghold_password(self):
        """Clear stronghold password.
        """
        return self._send_cmd_routine(
            'clearStrongholdPassword'
        )

    def is_stronghold_password_available(self):
        """Is stronghold password available.
        """
        return self._send_cmd_routine(
            'isStrongholdPasswordAvailable'
        )

    def recover_accounts(self, account_start_index, account_gap_limit, address_gap_limit, sync_options):
        """Recover accounts.
        """
        return self._send_cmd_routine(
            'recoverAccounts', {
                'accountStartIndex': account_start_index,
                'accountGapLimit': account_gap_limit,
                'addressGapLimit': address_gap_limit,
                'syncOptions': sync_options
            }
        )

    def remove_latest_account(self):
        """Remove latest account.
        """
        return self._send_cmd_routine(
            'removeLatestAccount'
        )

    def restore_backup(self, source, password):
        """Restore a backup from a Stronghold file
           Replaces client_options, coin_type, secret_manager and accounts. Returns an error if accounts were already created
           If Stronghold is used as secret_manager, the existing Stronghold file will be overwritten. If a mnemonic was
           stored, it will be gone.
        """
        return self._send_cmd_routine(
            'restoreBackup', {
                'source': source,
                'password': password
            }
        )

    def generate_mnemonic(self):
        """Generates a new mnemonic.
        """
        return self._send_cmd_routine(
            'generateMnemonic'
        )

    def verify_mnemonic(self, mnemonic):
        """Checks if the given mnemonic is valid.
        """
        return self._send_cmd_routine(
            'verifyMnemonic', {
                'mnemonic': mnemonic
            }
        )

    def set_client_options(self, client_options):
        """Updates the client options for all accounts.
        """
        return self._send_cmd_routine(
            'setClientOptions',
            {
                'clientOptions': client_options
            }
        )

    def generate_address(self, account_index, internal, address_index, options=None, bech32_hrp=None):
        """Generate an address without storing it.
        """
        return self._send_cmd_routine(
            'generateAddress', {
                'accountIndex': account_index,
                'internal': internal,
                'addressIndex': address_index,
                'options': options,
                'bech32Hrp': bech32_hrp
            }
        )

    def get_node_info(self, url, auth):
        """Get node info.
        """
        return self._send_cmd_routine(
            'getNodeInfo', {
                'url': url,
                'auth': auth
            }
        )

    def set_stronghold_password(self, password):
        """Set stronghold password.
        """
        return self._send_cmd_routine(
            'setStrongholdPassword', {
                'password': password
            }

        )

    def set_stronghold_password_clear_interval(self, interval_in_milliseconds):
        """Set stronghold password clear interval.
        """
        return self._send_cmd_routine(
            'setStrongholdPasswordClearInterval', {
                'intervalInMilliseconds': interval_in_milliseconds
            }
        )

    def store_mnemonic(self, mnemonic):
        """Store mnemonic.
        """
        return self._send_cmd_routine(
            'storeMnemonic', {
                'mnemonic': mnemonic
            }

        )

    def start_background_sync(self, options, interval_in_milliseconds):
        """Start background sync.
        """
        return self._send_cmd_routine(
            'startBackgroundSync', {
                'options': options,
                'intervalInMilliseconds': interval_in_milliseconds
            }
        )

    def stop_background_sync(self):
        """Stop background syncing.
        """
        return self._send_cmd_routine(
            'stopBackgroundSync',
        )

    def listen(self, handler, events=None):
        """Listen to wallet events, empty array or None will listen to all events
           The default value for events is None
        """
        events_array = [] if events is None else events
        iota_wallet.listen(self.handle, events_array, handler)

    def clear_listeners(self, events=None):
        """Remove wallet event listeners, empty array or None will remove all listeners
           The default value for events is None
        """
        events_array = [] if events is None else events
        return self._send_cmd_routine(
            'clearListeners', {
                'eventTypes': events_array
            }

        )

    @staticmethod
    def __return_str_or_none(str):
        if str:
            return str
        else:
            return None
